src/main/java/jenkins/plugins/logstash/persistence/LogzioDao.java

```diff
@@ -0,0 +1,127 @@
+    private final LogzioSender logzioSender;
+
+    private String key;
+    private String host;
+
+    //primary constructor used by indexer factory
+    public LogzioDao(String host, String key){
+        this(null, host, key);
+    }
+
+    // Factored for unit testing
+    LogzioDao(LogzioSender factory, String host, String key){
+        this.host = host;
+        this.key = key;
+
+        // create file for sender queue
+        File fp;
+        Jenkins jenkins = Jenkins.getInstanceOrNull();
+        if (jenkins == null){
+            // for testing (mvn package)
+            fp = Files.createTempDir();
+        }else{
+            fp = new File(jenkins.getRootDir() + "/logs/logzio_plugin");
+            boolean folderExisted = fp.mkdirs() || fp.setWritable(true);
+            if (!folderExisted) {
+                throw new IllegalArgumentException("Unable to create path");
+            }
+        }
+
+        try{
+            this.logzioSender = factory == null ? LogzioSender.getOrCreateSenderByType(key, TYPE, DRAIN_TIMEOUT,
+                    FS_PERCENT_THRESHOLD, fp, host, SOCKET_TIMEOUT, CONNECT_TIMEOUT,false, null,
+                    Executors.newScheduledThreadPool(CORE_POOL_SIZE),GC_PERSISTED_QUEUE_FILE_INTERVAL_SECOND, true) : factory;
+            this.logzioSender.start();
+        }catch (LogzioParameterErrorException e){
+            throw new IllegalArgumentException(e.getMessage());
+        }
+    }
+
+    @Override
+    public void push(String data) {
+        JSONObject jsonData = JSONObject.fromObject(data);
+        JSONArray logMessages = jsonData.getJSONArray("message");
+        for (Object logMsg : logMessages) {
+            JsonObject logLine = createLogLine(jsonData, logMsg.toString());
+            this.logzioSender.send(logLine);
+        }
+    }
+
+    protected JsonObject createLogLine(JSONObject jsonData, String logMsg) {
+        JsonObject logLine = new JsonObject();
+        logLine.addProperty("message", logMsg);
+        logLine.addProperty("@timestamp", LogstashConfiguration.getInstance().getDateFormatter().format(Calendar.getInstance().getTime()));
+        for(Object key : jsonData.keySet()){
+            String keyStr = (String) key;
+            if (!keyStr.equals("message")){
+                logLine.addProperty(key.toString(), jsonData.getString(key.toString()));
+            }
+        }
+        return logLine;
+    }
+
+    @Override
+    public JSONObject buildPayload(BuildData buildData, String jenkinsUrl, List<String> logLines) {
+        JSONObject payload = new JSONObject();
+        payload.put("message", logLines);
+        payload.put("source", "jenkins");
+        payload.put("source_host", jenkinsUrl);
+        payload.put("@buildTimestamp", buildData.getTimestamp());
+        payload.put("@version", 1);
+        // flatten build data
+        Map<String, Object> flattenJson = JsonFlattener.flattenAsMap(buildData.toString());
```

Comment:
so logz.io is based on Elasticsearch obviously. #21 also addresses the problem of keys containing illegal characters. Maybe one could have a common helper for both.

Reply:
Kibana has problems with '.' if I remember correct, better to use '_'

Reply:
that is exactly addressed by #21 and it is ES that has this problem afaik

Reply:
Where do you think the common helper should be?

Reply:
@jakub-bochenski  any news?

Reply:
on second thoughts lets leave it here for now (add a TODO marker)
the logic in #21 is only changing the env variable names while this is flattening all the entires

Reply:
please add comment explaining the need to flatten data (best link to logz.io docs about input format)